docs/developer/web.rst

```diff
@@ -0,0 +1,224 @@
+    the allowed methods. This method **may** be combined with the following two
+    request headers:
+
+    - ``Access-Control-Request-Headers`` `(more on request headers)
+      <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Request-Headers>`_
+    - ``Access-Control-Request-Method`` `(more on request methods)
+      <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Request-Method>`_
+
+**Head**
+
+    ``HEAD`` method can be used to request the headers that would be returned
+    if a given resource were requested using ``GET``. The response to ``HEAD``
+    **must not** include a body.
+
+In case you need more verbs or you are not working with a resource, you should
+pause since you might be designing a new **communication protocol**.
+
+.. _`CRUD`: https://en.wikipedia.org/wiki/Create,_read,_update_and_delete
+
+3. Content Negotiation
+----------------------
+
+``Accept``
+    All meta-data endpoints should support ``application/json`` and
+    optionally they may support ``application/vnd.api+json`` if they
+    conform with `JSON API specification`_.
+
+
+4. Errors
+---------
+
+In case of an error, the ``Content-Type`` of the response **must** match the
+``Accept`` header of the request. A JSON error response **must** contain an
+error object in the response body. The error object **should** contain a unique
+error code (that can be looked up for more details), an error message and
+possibly a detailed description:
+
+.. code-block:: JSON
+
+    {
+        "code": 1234,
+        "message": "Something went wrong.",
+        "description": "What exactly went wrong."
+    }
+
+Furthermore, every error response **must** use a meaningful `HTTP status code
+<https://tools.ietf.org/html/rfc7231#section-6.1>`_.
+The status code **may** also be part of the error object in the response body.
+
+
+5. URL structure
+----------------
+
+URLs **should** be RESTful in the sense that they represent logical resources
+or collections of resources. The plural form should be used as endpoint names.
+Whenever possible, a single word should be used to describe resources. Where
+compound words cannot be avoided, they are built using **dashes**. For complex
+queries (filtering, sorting) query parameters **should** be used.
+
+
+- URL for the entire collection of contexts:
+    ``/contexts``
+
+- URL for a specific context:
+    ``/contexts/{id}``
+
+
+5.1 Relations
+~~~~~~~~~~~~~
+A relation between resources **must** be represented in the URL structure if
+the related resources only exist in the context of the relation. For example,
+if in a situation where a file has primary key composed of a bucket identifier
+and a file identifier, the URLs would be structured as followed:
+
+- URL for all files inside the bucket:
+    ``/buckets/{bucket}/files``
+
+- URL for a specific file in a given bucket:
+    ``/buckets/{bucket}/files/{file}``
```

Comment:
As a file can technically be replicated in several buckets, would this one be also a redirect to something like /files/{file} like in the example below ?

Reply:
I don't think you can access a file outside a context of bucket.

Reply:
Buckets exist for administrative and technical reasons. Files are conceptually not dependent on buckets.

Reply:
Feel free to find another example, however I still think that the only way how you can access file is via a bucket.

Reply:
A file exists independently of its actual storage and you may want to refer to it in a general way (just with its UUID for example), then when actually reading the file it would be dereferenced to the most suited location, like closest to the current execution (that's the reason why we have the file_location nodes which are themselves necessarily linked to a bucket). This allows also for referencing files that are outside of the platform and which would have a sort of "external ref" node instead of the file_location. 

You can think about a "file" as a conceptual dataset (like the well-known iris dataset), file_location is the physical instance of this dataset stored somewhere (like iris.dat in the bucket "my-data" on SWITCH), you also have file_versions which represents a specific version of this file on the specific storage (which could be using S3 versioning system, or just adding the timestamp to the filename as we do for local storage).

Reply:
I don't think the discussion about external files should be linked to URL structure. The point here was that if we have files attached to bucket that can have an access control, then we need to make sure that the file is accessed via that bucket. Hence, the developer should structure the URL as described.

Personally, I would like to separate the discussion about protocol/API to access file and the resource definition.

Reply:
So I guess in renga we don't have a really good example for the case where a resource exists only in relation to another resource then. I suggest I just use a made-up non-renga example here instead.

Reply:
I agree that the discussion of the representation of the specific resource of type "file" is not the main point I wanted to bring here. I was just saying that depending on how it is dealt with later on, taking the example of the file may lead to misunderstanding of the users. As most of our resources will need some very unique identifiers, they may all end up being first-class citizens that can be accessed directly.
As I have no better example of a resource that is meaningless without its relation with another one, we can either keep it like it is now until the resource representation is clearly defined or take a fake example to convey our message, which is still relevant in this document.